lib/backends/webgl/inference-handler.ts

```diff
@@ -5,105 +5,151 @@ import {InferenceHandler} from '../../backend';
+   * @param tensor the tensor with data to upload
+   */
+  getOrCreateTextureData(tensor: Tensor, layout?: TextureLayout) {
+    let td = this.getTextureData(tensor.dataId);
     if (!td) {
       Logger.verbose('InferenceHandler', `Creating new TextureData for dims: [${tensor.dims}]`);
       if (!layout) {
-        layout = this.createBasicTextureLayout(tensor.dims.slice());
+        layout = this.createTextureLayoutFromShape(tensor.dims.slice());
       }
       // graph inputs or initializers
-      td = this.createTextureDataFromLayout(layout, tensor.type, tensor.numberData, Encoder.Usage.UploadOnly);
-      this.setTextureData(tensor, td);
+      td = this.createTextureData(layout, tensor.type, tensor.numberData, tensor, Encoder.Usage.UploadOnly);
     } else {
       Logger.verbose('InferenceHandler', `Retrieving TextureData from cache: [${tensor.dims}]`);
     }
     return td;
   }
-  getTextureData(tensor: Tensor): TextureData|undefined {
-    return this.lookupTextureData(tensor);
+
+  /**
+   * Create a TextureData object from the given data type and texture layout.
+   * Usage = Encoder.Usage.Default.
+   * @param dataType the tensor data type
+   */
+  createTextureDataFromLayout(layout: TextureLayout, dataType: Tensor.DataType): TextureData {
+    return this.createTextureData(layout, dataType);
   }
-  setTextureData(tensor: Tensor, td: TextureData): void {
-    if (this.session.isInitializer(tensor)) {
-      this.session.setTextureData(tensor, td);
-      return;
-    }
-    this.tensorToTexture.set(tensor, td);
-    this.textureToTensor.set(td, tensor);
+
+  /**
+   * Create a TextureData object using the given data and bind to the given tensor.
+   * Usage = Encoder.Usage.UploadOnly.
+   * NOTE: this function is a hack for Conv implementation. should remove this function, after rewriting Conv
+   * implementation by Graph.Transformer
+   * @param dataType the tensor data type
+   * @param data the actual data to upload
+   * @param tensor the tensor to bind. tensor's data is ignored.
+   */
+  createTextureDataFromLayoutBindTensor(
+      layout: TextureLayout, dataType: Tensor.DataType, data: Tensor.NumberType, tensor: Tensor): TextureData {
+    return this.createTextureData(layout, dataType, data, tensor, Encoder.Usage.UploadOnly);
+  }
+
+  private createTextureData(
+      layout: TextureLayout, dataType: Tensor.DataType, data?: Tensor.NumberType, tensor?: Tensor,
+      usage?: Encoder.Usage): TextureData {
+    Logger.verbose('InferenceHandler', `Creating TextureData: layout:[${JSON.stringify(layout)}]`);
+    const texture = this.textureHelper.createTextureFromLayout(dataType, layout, data, usage);
+    return this.createTextureDataFromTexture(layout, dataType, texture, tensor);
+  }
+
+  /**
+   * Create a TextureData object, using the given texture.
+   * This function does not create new texture. Usually used in scenarios using texture sharing. (eg. Reshape)
```

Comment:
Can you expand on `texture sharing` for my understanding please ?

Reply:
A `WebGLTexture` object represents a texture in GPU.

A `TextureData` object represents a corresponding texture to a tensor. It contains a `WebGLTexture` object, a reference to the tensor and the texture layout.

A `Tensor.Id` object represents a unique data unit, regardless where the data is hold (it can be in memory, ie. TypedArray, or in GPU, ie. a texture).

A `Tensor` object represent a value node in the model graph. It's one of those:
  - a model input
  - a model initializer
  - a model output
  - an intermedia value through the model execution
A `Tensor` object can have multiple way to connect to the data that its tensorId mapped to:
  - a cached result (TypedArray instance)
  - a data provider function ( a function that returns a TypedArray instance )
  - an async data provider function (this is not supported yet)

Logically, `WebGLTexture` objects should be 1:1 mapping to `Tensor.Id` objects, and `TextureData` objects should be a 1:1 mapping to `Tensor` objects.

Some operators, for example, `Reshape`, will not modify any data of its input. In this case, we are creating a new `Tensor` object with different `dims`, using the same `Tensor.Id`. This is the sharing scenario. We extend this scenario with the equivalent concepts in webgl.